### PR TITLE
[ FSU ] Enabls Asynchronos FSU for forwarding

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR enables asynchronos mode for FSU (flash storage utilization)
for better performance.

It splits the load and unload tensors which make difficult to handle.
Also fix the inference execution order when it is in INFERENCE mode
and change the trainable option to false when it calls the request
weights and tensors.

Add the new function to load and unload tensors as well as check load
complete.

It also considers weight pool and tensor pool differenetly according
to the ExecutionMode. It is not use FSU mode for tensor pool for the
INFERENCE Mode.

Resolves:

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: Applications/SimpleFC/jni/main.cpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
```
Comment (baek2sm) [inline Applications/SimpleFC/jni/main.cpp:3]: name and date?

**Review by baek2sm — APPROVED** [review]
LGTM

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Review by djeong20 — APPROVED** [review]
Appreciate the hard work!

File: nntrainer/tensor/manager.cpp [hunk 2]
```diff
@@ -730,6 +738,117 @@ void Manager::flushCache() {
...
+        t_fut.wait();
+    } else {
+      auto w_fut = completed_load_tensor[std::get<0>(tasks)].get_future();
+      lock.unlock();
+      if (std::get<0>(tasks) != 0)
+        w_fut.wait();
+    }
+    async_load_tensor.erase(order);
+    ml_logd("wait and completed %d", order);
+    ;
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/tensor/manager.cpp:761]: thansk. fixed.

File: nntrainer/models/neuralnet.cpp [hunk 3]
```diff
@@ -336,9 +351,38 @@ sharedConstTensors NeuralNetwork::forwarding(
...
+        unsigned int lookahead =
+          std::get<props::MemorySwapLookahead>(model_flex_props);
+
+        if (lookahead != 0) {
+          if ((f) % (lookahead + 1) == lookahead - 1) {
+            std::cout << "request load tensor : " << f + lookahead + 1
+                      << std::endl;
+            ml_logd("request load tensor for %d", f + 1);
+            model_graph.LoadTensors((f / (lookahead + 1) + 1) *
+                                    (lookahead + 1));
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/models/neuralnet.cpp:376]: I left the comments in code.

File: nntrainer/models/neuralnet.cpp [hunk 4]
```diff
@@ -214,6 +213,20 @@ int NeuralNetwork::compile(ExecutionMode mode) {
...
 
+  if (mode != exec_mode) {
+    if (mode == ExecutionMode::INFERENCE) {
+      ml_logd("Execution mode mismatch : train mode @compile & inference mode "
+              "@ initialize");
+      exec_mode = mode;
+    } else {
+      NNTR_THROW_IF(((mode == ExecutionMode::INFERENCE) &&
+                     (exec_mode == ExecutionMode::TRAIN)),
+                    std::invalid_argument)
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/models/neuralnet.cpp:224]: fixed. Thanks.

File: Applications/SimpleFC/jni/meson.build [hunk 5]
```diff
@@ -0,0 +1,28 @@
+resnet_sources = [
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply Applications/SimpleFC/jni/meson.build:1]: fixed. Thanks.

File: Applications/SimpleFC/jni/main.cpp [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  for (; iter != value.end() - 1; ++iter) {
+    ss << *iter << ',';
+  }
+  ss << *iter;
+
+  return ss.str();
+}
+
+/**
+ * @brief Create resnet 18
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply Applications/SimpleFC/jni/main.cpp:79]: fixed. Thanks.

File: Applications/SimpleFC/jni/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply Applications/SimpleFC/jni/main.cpp:3]: fixed. Thanks.

Comment (jijoongmoon) [reply Applications/SimpleFC/jni/main.cpp:3]: fixed. Thanks.

File: nntrainer/models/neuralnet.cpp [hunk 7]
```diff
@@ -329,16 +342,73 @@ NeuralNetwork::~NeuralNetwork() {
...
+
+       Step 3. Then check the look a head which says how many layer weights need
+               to be loaded before running to hide overehad due to FSU,
+
+       Step 4. Try to get the tesors by asking tensors for layers which is done
+               by thread pool
+
+       Step 5. Try to release the weights which has execution order less then f.
+
+       Step n. repeat next layer starting with checking the tenosrs are loaded,
```
Comment (SeoHyungjun) [inline nntrainer/models/neuralnet.cpp:381]: Is this a typo?

```suggestion
       Step n. repeat next layer starting with checking the tensors are loaded,
```

**Pull request merged**